imap_processing/hit/l0/decom_hit.py

```diff
@@ -0,0 +1,485 @@
+        #  - subcommutate sectorates
+        #  - decompress data
+        #  - Follow up with HIT team about erates and evrates.
+        #    (i.e.Should these be arrays containing all the sub fields
+        #    or should each subfield be it's own data field/array)
+
+        # Get dims for data variables (yaml file not created yet)
+        if len(field_meta.shape) > 1:
+            dims = ["epoch", "gain", f"{field}_index"]
+        elif field_meta.shape[0] > 1:
+            dims = ["epoch", f"{field}_index"]
+        else:
+            dims = ["epoch"]
+
+        sci_dataset[field] = xr.DataArray(parsed_data, dims=dims, name=field)
+        # increment the start of the next section of data to parse
+        section_start += field_meta.section_length
+
+
+def is_sequential(counters: np.ndarray) -> np.bool_:
+    """
+    Check if an array of packet sequence counters is sequential.
+
+    Parameters
+    ----------
+    counters : np.ndarray
+        Array of packet sequence counters.
+
+    Returns
+    -------
+    bool
+        True if the sequence counters are sequential, False otherwise.
+    """
+    return np.all(np.diff(counters) == 1)
+
+
+def find_valid_starting_indices(flags: np.ndarray, counters: np.ndarray) -> np.ndarray:
+    """
+    Find valid starting indices for science frames.
+
+    This function finds the starting indices of valid science frames.
+    A valid science frame has the following packet grouping flags:
+
+            First packet: 1
+            Next 18 packets: 0
+            Last packet: 2
+
+    The packet sequence counters for the identified science frames must
+    be sequential. Only the starting indices of valid science frames are
+    returned.
+
+    Parameters
+    ----------
+    flags : np.ndarray
+        Array of packet grouping flags.
+    counters : np.ndarray
+        Array of packet sequence counters.
+
+    Returns
+    -------
+    valid_indices : np.ndarray
+        Array of valid indices for science frames.
+    """
+    # Define the pattern of grouping flags in a complete science frame.
+    flag_pattern = np.array(
+        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]
+    )
+    # Use sliding windows to compare segments of the array (20 packets) with the
+    # pattern. This generates an array of overlapping sub-arrays, each of length
+    # 20, from the flags array and is used to slide the "window" across the array
+    # and compare the sub-arrays with the predefined pattern.
+    window_size = len(flag_pattern)
+    windows = np.lib.stride_tricks.sliding_window_view(flags, window_size)
+    # Find where the windows match the pattern
+    matches = np.all(windows == flag_pattern, axis=1)
+    # Get the starting indices of matches
+    match_indices = np.where(matches)[0]
+    # Filter for only indices from valid science frames with sequential counters
+    valid_indices = get_valid_indices(match_indices, counters, window_size)
```

Comment:
I wonder if you can combine remaining two functions into one to reduce code tracing. You can ignore for now but something to think about going forward.

Reply:
Great idea. I've added a TODO for now so I don't forget